runner/scripts/3-runAssignments.ts

```diff
@@ -1,45 +1,134 @@
+    case "sa": {
+      return RunnerID.SecretAgent;
+    }
+
+    case "puppeteer":
+    case "pptr": {
+      return RunnerID.Puppeteer;
+    }
+
+    default: {
+      console.warn(`parseRunnerID: ignore unrecognized runner value: '${value}'`);
+      return previous;
+    }
+  }
+}
+
+function parseNumber(value: string): number {
+  return parseInt(value);
+}
+
+program
+  .option('-r, --runner <hero|sa|secret-agent|pptr|puppeteer>', 'select the runner to run', parseRunnerID, RunnerID.Hero)
+  .option('--secret-agent-port <port>', 'select port to use for secret agent (flag only used if secret agent runner is selected)', parseNumber, 7007);
+program.parse();
+const options = program.opts();
 
 // process.env.SA_SHOW_BROWSER = 'true';
 process.env.SA_SHOW_REPLAY = 'false';
 
 const TYPE = 'external';
 
-(async function run() {
-  await Core.start();
-
-  const runAssignment = async (assignment: IAssignment) => {
-    const agentMeta = RealUserAgents.extractMetaFromUserAgentId(assignment.userAgentId);
-
-    const agent = new Hero({
-      userAgent: `~ ${agentMeta.operatingSystemName} = ${agentMeta.operatingSystemVersion.replace(
-        '-',
-        '.',
-      )} & ${agentMeta.browserName} = ${agentMeta.browserVersion.replace('-0', '')}`,
-    });
-    console.log(assignment.userAgentString);
-    console.log('AGENT: ', await agent.meta);
-    await runAssignmentInHero(agent as any, assignment);
-    await agent.close();
-  };
-  const userAgentsToTestPath = Path.join(
-    __dirname,
-    `../data/${TYPE}/2-user-agents-to-test/userAgentsToTest`,
-  );
+async function runFactoryRunners(runnerID: RunnerID, runnerFactory: IRunnerFactory) {
+  console.log(`run all assignments for runner: ${runnerID}!`);
 
+  const userAgentsToTestPath = Path.join(__dirname, `../data/${TYPE}/2-user-agents-to-test/userAgentsToTest`);
   const config = {
-    userId: 'testing',
+    userId: `runner-${runnerID}`,
     dataDir: Path.resolve(__dirname, `../data/${TYPE}/3-assignments`),
     concurrency: 1,
     userAgentsToTestPath,
   };
-  await forEachAssignment(config, assignment => runAssignment(assignment));
-})()
-  .then(() => process.exit())
-  .catch(console.log);
+
+  await forEachAssignment(config, async (assignment) => {
+    let runner: IRunner;
+    try {
+      runner = await runnerFactory.spawnRunner(assignment);
+    } catch (error) {
+      console.error(`failed to create runner ${runnerID}: ${error}`);
+      return;
+    }
+    try {
+      await runner.run(assignment);
+    } catch (error) {
+      console.error(`runner ${runnerID} run failed with exception: ${error}`);
+    } finally {
+      try {
+        await runner.stop();
+      } catch (error) {
+        console.error(`failed to stop runner ${runnerID}: ${error}`);
+      }
+    }
+  });
+}
+
+async function run() {
+  const runnerId = options.runner || RunnerID.Puppeteer;
+  let runnerFactory: IRunnerFactory;
+
+  switch (runnerId) {
+    case RunnerID.Puppeteer: {
```

Comment:
Since this is heading towards a plugin structure, it might be interesting to think of a way for each Factory to register itself into this tool so that we can extend it with custom runner factories. Perhaps a future idea.